File: backend/src/main/java/dev/coldhands/pair/stairs/backend/infrastructure/web/controller/CurrentUserController.java
package dev.coldhands.pair.stairs.backend.infrastructure.web.controller;

import dev.coldhands.pair.stairs.backend.domain.User;
import dev.coldhands.pair.stairs.backend.infrastructure.persistance.repository.UserRepository;
import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.CurrentUserDto;
import dev.coldhands.pair.stairs.backend.usecase.UserDetailsService;
import org.springframework.boot.autoconfigure.condition.ConditionalOnBooleanProperty;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.security.oauth2.core.oidc.user.OidcUser;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@ConditionalOnBooleanProperty("app.feature.flag.teams.enabled")
public class CurrentUserController {

    private final UserDetailsService userDetailsService;

    public CurrentUserController(UserDetailsService userDetailsService) {
        this.userDetailsService = userDetailsService;
    }

    @GetMapping("/api/v1/me")
    public CurrentUserDto getCurrentUser(@AuthenticationPrincipal OidcUser oidcUser) {
        final var userInfo = oidcUser.getUserInfo();

        final var user = userDetailsService.getUserByOidcSub(userInfo.getSubject());

        return new CurrentUserDto(userInfo.getFullName(), user.displayName());
    }
}

File: backend/src/main/kotlin/dev/coldhands/pair/stairs/backend/infrastructure/web/controller/CurrentUserController.kt
package dev.coldhands.pair.stairs.backend.infrastructure.web.controller

import dev.coldhands.pair.stairs.backend.domain.OidcSub
import dev.coldhands.pair.stairs.backend.infrastructure.web.dto.CurrentUserDto
import dev.coldhands.pair.stairs.backend.usecase.UserDetailsService
import org.springframework.boot.autoconfigure.condition.ConditionalOnBooleanProperty
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.security.oauth2.core.oidc.user.OidcUser
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@ConditionalOnBooleanProperty("app.feature.flag.teams.enabled")
class CurrentUserController(private val userDetailsService: UserDetailsService) {
    @GetMapping("/api/v1/me")
    fun getCurrentUser(@AuthenticationPrincipal oidcUser: OidcUser): ResponseEntity<CurrentUserDto> {
        val userInfo = oidcUser.userInfo

        return userDetailsService.getUserByOidcSub(OidcSub(userInfo.subject))
            ?.let { ResponseEntity.ok(CurrentUserDto(userInfo.fullName, it.displayName)) }
            ?: ResponseEntity.notFound().build()
    }
}